src/main/webapp/app/grading-system/grading-key-overview/grading-key-overview.component.ts

```diff
@@ -42,35 +42,53 @@ export class GradingKeyOverviewComponent implements OnInit {
-                            const maxPoints = this.courseCalculationService.calculateTotalScores(course!.exercises!, course!).get(ScoreType.REACHABLE_POINTS);
-                            this.gradingSystemService.setGradePoints(this.gradeSteps, maxPoints!);
-                        } else {
-                            // for exams the max points filed should equal the total max points (otherwise exams can't be started)
-                            this.gradingSystemService.setGradePoints(this.gradeSteps, gradeSteps.maxPoints!);
-                        }
+        // Note: This component is used in multiple routes, so it can be lazy loaded. Also, courseId and examId can be
+        // found on different levels of hierarchy tree (on the same level or a parent or a grandparent, etc.).
+        this.courseId = Number(this.findParamInRouteHierarchy('courseId'));
+        const examIdParam = this.findParamInRouteHierarchy('examId');
+        if (examIdParam) {
+            this.examId = Number(examIdParam);
+            this.isExam = true;
+        }
+        this.gradingSystemService.findGradeSteps(this.courseId, this.examId).subscribe((gradeSteps) => {
+            if (gradeSteps) {
+                this.title = gradeSteps.title;
+                this.isBonus = gradeSteps.gradeType === GradeType.BONUS;
+                this.gradeSteps = this.gradingSystemService.sortGradeSteps(gradeSteps.gradeSteps);
+                if (gradeSteps.maxPoints !== undefined) {
+                    if (!this.isExam) {
+                        // calculate course max points based on exercises
+                        const course = this.courseCalculationService.getCourse(this.courseId!);
+                        const maxPoints = this.courseCalculationService.calculateTotalScores(course!.exercises!, course!).get(ScoreType.REACHABLE_POINTS);
+                        this.gradingSystemService.setGradePoints(this.gradeSteps, maxPoints!);
+                    } else {
+                        // for exams the max points filed should equal the total max points (otherwise exams can't be started)
+                        this.gradingSystemService.setGradePoints(this.gradeSteps, gradeSteps.maxPoints!);
                     }
                 }
-            });
-        });
-        this.route.parent?.parent?.queryParams.subscribe((queryParams) => {
-            this.studentGrade = queryParams['grade'];
+            }
         });
+
+        // Needed queryParam is available on this component so no need to traverse the hierarchy like params above.
+        this.studentGrade = this.route.snapshot.queryParams['grade'];
+    }
+
+    /**
+     * Checks router hierarchy to find a given paramKey, starting from the current ActivatedRouteSnapshot
+     * and traversing the parents.
+     * @param paramKey the desired key of route.snapshot.params
+     * @private
+     */
+    private findParamInRouteHierarchy(paramKey: string): string | undefined {
```

Comment:
💡Idea: It would be very cool if you could extract this into a service, because we could use it in different places. Similiar to how it is done here: https://gist.github.com/jessemorton/3ea95bce1caae0234235f72530cbb45e

Reply:
Sounds like a good idea! I think we can address it in a follow-up when the bugfix is merged. Do you have an existing service  in mind that we can use or do we go with a new one?

Reply:
I think a new one like the example would be nice :)

Reply:
Good suggestion, but rather than a service I would move this to one of your helpers in the `utils` directory. 🙂